src/analyses/dirty.cpp

```diff
@@ -19,14 +19,49 @@ Date: March 2013
 void dirtyt::build(const goto_functiont &goto_function)
 {
   for(const auto &i : goto_function.body.instructions)
-    i.apply([this](const exprt &e) { find_dirty(e); });
+  {
+    if(i.is_other())
+    {
+      search_other(i);
+    }
+    else
+    {
+      i.apply([this](const exprt &e) { find_dirty(e); });
+    }
+  }
+}
+
+void dirtyt::search_other(const goto_programt::instructiont &instruction)
+{
+  INVARIANT(instruction.is_other(), "instruction type must be OTHER");
+  if(instruction.get_other().id() == ID_code)
+  {
+    const codet &code = instruction.get_other();
+    const irep_idt &statement = code.get_statement();
+    if(
+      statement == ID_expression || statement == ID_array_set ||
+      statement == ID_array_equal || statement == ID_array_copy ||
+      statement == ID_array_replace || statement == ID_havoc_object ||
+      statement == ID_input || statement == ID_output)
+    {
+      forall_operands(it, code)
+        find_dirty(*it);
+    }
+    // other possible cases according to goto_programt::instructiont::output
+    // and goto_symext::symex_other:
+    // statement == ID_fence ||
+    // statement == ID_user_specified_predicate ||
+    // statement == ID_user_specified_parameter_predicates ||
+    // statement == ID_user_specified_return_predicates ||
+    // statement == ID_decl || statement == ID_nondet || statement == ID_asm)
+  }
 }
 
 void dirtyt::find_dirty(const exprt &expr)
 {
-  if(expr.id()==ID_address_of)
+  if(expr.id() == ID_address_of)
```

Comment:
Don't change it on this PR but in general it is good to have refactoring in a separate commit to make reviewing and bisecting for regressions easier.